### PR TITLE
Alerting: Update state manager to change all current states in the case when Error\NoData is executed as Ok\Nomal

### LINKED ISSUE #66790: Incorrect mapping of NoData\Error to multidimensional rules
## What happened:
Execution of NoData\Error set to Ok or Alerting of multi-dimensional rules creates a new alert instance instead of setting the state of existing instances. 

<details><summary>Example</summary>

https://user-images.githubusercontent.com/25988953/232835917-853cc210-4dc0-4886-a8f7-cdcf8e2bb691.mp4

</details>

## What you expected to happen

State of the current alert rule instances should be set according to the mapping options.

## How to reproduce it (as minimally and precisely as possible):

- Create a rule that returns multi-dimensional results
- Wait util it is evaluated.
- Make data source provide either NoData or Error 
- Wait for the next evaluation
- See that a new state is created.

## Environment:
- Grafana version: main-4a7ecc24
- Data source type & version: N/A
- OS Grafana is installed on: N/A
- User OS & Browser: N/A
- Grafana plugins: N/A
- Others: N/A

### PR DESCRIPTION
**Pull request by yuri-tceretian**
**What is this feature?**
This PR changes how executions of NoData\Error as OK\Alerting are handled by alerting the state manager. In those cases, the state manager updates all current states according to the execution settings.

<details><summary>Old Behavior</summary>

https://user-images.githubusercontent.com/25988953/232835917-853cc210-4dc0-4886-a8f7-cdcf8e2bb691.mp4

</details>

<details><summary>New Behavior</summary>

https://github.com/grafana/grafana/assets/25988953/f1bc7461-0ed1-47a7-9c50-c4a726182ae5

</details>

The feature is behind the flag `alertingNoDataErrorExecution`

**Why do we need this feature?**
To fix execution of NoData\Error when it is set to Alerting\OK

**Who is this feature for?**
Alerting users who would like to treat exceptional states (NoData\Error) by maintaining the current state.

**Which issue(s) does this PR fix?**:

Fixes https://github.com/grafana/grafana/issues/66790

**Special notes for your reviewer:**

~Please review PR by commit. I copied test cases introduced in https://github.com/grafana/grafana/pull/73019 and fixed to pass when the feature flag is enabled. See commit https://github.com/grafana/grafana/pull/68142/commits/0ec9a026c84f16b0c91aa445dc124bf42bb1893e to understand the difference.~
After Matt's suggestion, I decided to amend to the existing tests instead of copying the whole suite. So, I updated the tests to run each test case twice - when flag is enabled and disabled. This shows that normal transitions are not affected, as well as Execution of NoData\Error as NoData\Error. The test cases where the difference exists, are overridden by a new set of assertions.
Tests that have overridden assertions are marked by `[*]`

<details><summary>Example</summary>

image

</details>

Please check that:
- [x] It works as expected from a user's perspective.
- [x] If this is a pre-GA feature, it is behind a feature toggle.
- [x] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

**Comment by JacobsonMT** [comment]
Instead of copying the tests, what do you think about something like this:

```GO
func TestProcessEvalResults_StateTransitions(t *testing.T) {
	t.Run("Without applyNoDataAndErrorToAllStates", func(t *testing.T) {
		stateTransitions(t, false)
	})
	t.Run("With applyNoDataAndErrorToAllStates", func(t *testing.T) {
		stateTransitions(t, true)
	})
}

func stateTransitions(t *testing.T, applyNoDataAndErrorToAllStates bool) {
...
}
```

Then making the differences explicit based on the value of `applyNoDataAndErrorToAllStates`.

I think this will make the PR much simpler to read and the tests should be less at risk of drift between feature flag true and false.

File: pkg/services/ngalert/state/manager.go [hunk 1]
```diff
@@ -288,7 +325,7 @@ func (st *Manager) setNextState(ctx context.Context, alertRule *ngModels.AlertRu
 	// Usually, it happens in the case of classic conditions when the evalResult does not have labels.
 	//
 	// This is temporary change to make sure that the labels are not persistent in the state after it was in Error state
-	// TODO yuri. Remove it in https://github.com/grafana/grafana/pull/68142
+	// TODO yuri. Remove it when correct Error result with labels is provided
```
Comment (yuri-tceretian) [inline pkg/services/ngalert/state/manager.go:328]: Can't remove it in this PR because it breaks execution of Error as Error.

**Review by JacobsonMT — APPROVED** [review]
LGTM, great job :rocket: 

image

**Review by grobinson-grafana — CHANGES_REQUESTED** [review]
I see this change is behind a feature flag, which is great! However, I'm requesting changes not because I want actual changes to be done to the code, but instead I want to ask a question before we add this to `main`.

tl;dr the question:

I understand the sentiment of this change is to make all states for a rule consistent when either No Data or an error occurs. What concerns me is that for some of our larger customers this can create massive alert storms. If their datasource is down, it could be 10,000s of alerts firing at once, perhaps 100,000s. If their grouping is generous, a number of contact points will also fail (such as email) as the email will be too large. I'm not sure about others such as Slack or Pagerduty.

I'm not sure this is the right feature to add moving forward, but I can support having it behind a feature flag for experimentation.

@yuri-tceretian and @JacobsonMT, could you share your opinions on what I've said above?

**Review by grobinson-grafana — APPROVED** [review]

**Comment by grobinson-grafana** [comment]
I tested a number of scenarios and seem to work as expected!
